File: Simulink/Code/FilterSample.m
function [output,delayline_out] = FilterSample(x,w,delayline_in);
% 
% Sample x is filtered with filter w. If x has multiple columns,
% each column of x is filtered with w. If w has multiple colums,
% the data x is filtered with both columns of w.
%
% Inputs:
%   x             = input data (Dimensions:1xnr_channels)
%   w             = time-domain filter coefficients (Dimensions:filterlength x nr_filters)
%   delayline_in  = input delayline (Dimensions:filterlength x nr_channels)
% Outputs:  
%   output        = output data (Dimensions: 1 x max(nr_channels,nr_filters))
%   delayline_out = output delayline
% 
%Date: December, 2007  
%Copyright: (c) 2007 by Ann Spriet
%e-mail: ann.spriet@esat.kuleuven.be   

tmp = size(x); % size  = 1 
nr_channels=tmp(2); % lấy số cột của [1 , 2]
tmp = size(w);
nr_filters=tmp(2); % lấy số cột  = 1 

if and(nr_filters>1,nr_channels>1);
  error('Nr_channels and nr_filters cannot be both larger than 1');
end

delayline_out = [x;delayline_in(1:end-1,:)];  % delay_out bằng giống với delaysample nó phải là một mảng 1 chiều 


if nr_channels>1
  output = w'*delayline_out; % w' là chuyển vị của w
else
  output = delayline_out'*w; % delayline_out' là chuyển vị của delayline_out * w : nhân ma trận thì ôutput phải là 1 giá trị 
end

File: Simulink/Code/De_tai_sinh_vien_2024-main/code_python/code/FilterSample.py
import numpy as np
from DelaySample import delay_sample


def filter_sample(x, w, delayline_in):
    """
    Lọc mẫu dữ liệu đầu vào x với bộ lọc w.
    
    Parameters:
    x : np.ndarray
        Dữ liệu đầu vào với kích thước (1, nr_channels).
    w : np.ndarray
        Các hệ số của bộ lọc thời gian với kích thước (filterlength, nr_filters).
    delayline_in : np.ndarray
        Hàng đợi mẫu đầu vào với kích thước (filterlength, max(nr_channels, nr_filters)).
    
    Returns:
    output : np.ndarray
        Dữ liệu đầu ra với kích thước (1, max(nr_channels, nr_filters)).
    delayline_out : np.ndarray
        Hàng đợi mẫu đầu ra với kích thước (filterlength, max(nr_channels, nr_filters)).
    """
    nr_channels = x.shape[1]
    nr_filters = w.shape[1]
    filterlength = w.shape[0]
    max_size = max(nr_channels, nr_filters)
    
    # Đảm bảo kích thước của x phù hợp với delayline_in
    if x.shape[0] != 1 or x.shape[1] != max_size:
        raise ValueError('x phải có kích thước (1, max(nr_channels, nr_filters))')
    
    # Cập nhật delayline_out
    delayline_out = np.vstack((x, delayline_in[:-1, :]))
    
    # Lọc mẫu x với bộ lọc w
    if nr_channels > 1:
        output = np.dot(w.T, delayline_out)
    else:
        output = np.dot(delayline_out.T, w)
    
    return output, delayline_out

